### PR TITLE
Add clamp conversion functionality

### PR DESCRIPTION
**Pull request by peri044**
# Description

Add clamp conversion

Fixes https://github.com/NVIDIA/TRTorch/issues/274

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/conversion/converters/impl/element_wise.cpp [hunk 1]
```diff
@@ -144,6 +144,41 @@ auto element_wise_registrations TRTORCH_UNUSED =
                     LOG_DEBUG("Output tensor shape: " << out->getDimensions());
                     return true;
                   }})
+        .pattern({"aten::clamp(Tensor self, Scalar? min=None, Scalar? max=None) -> (Tensor)",
```
Comment (narendasan) [inline core/conversion/converters/impl/element_wise.cpp:147]: Could we possibly lower to hardtanh or is the functionality different?

File: core/conversion/converters/impl/element_wise.cpp [hunk 2]
```diff
@@ -144,6 +144,41 @@ auto element_wise_registrations TRTORCH_UNUSED =
...
+                  [](ConversionCtx* ctx, const torch::jit::Node* n, args& args) -> bool {
+                    // Compute min(max(min_threshold, input), max_threshold)
+                    auto self = args[0].ITensorOrFreeze(ctx);
+                    auto clamp_layer_out = self;
+                    if (args[1].isIValue() && args[1].IValue()->isScalar()) {
```
Comment (narendasan) [inline core/conversion/converters/impl/element_wise.cpp:152]: I think we might be able to do !isNone on the IValue instead of two checks

Comment (peri044) [reply core/conversion/converters/impl/element_wise.cpp:152]: If I remember correctly, isNone() didn't work for me. I will double check again.

Comment (peri044) [reply core/conversion/converters/impl/element_wise.cpp:152]: The op does basically this `min(max(min_threshold, input), max_threshold)`
In the case when min_threshold or max_threshold is not given, args[1] or args[2] will be registered as an IValue. 
The isNone() check will check if the type is one of { kITensor, kIValue, kNone }. The type in this case would be IValue so isNone() will not return true.

File: core/conversion/converters/impl/element_wise.cpp (hunk 1 above)
Comment (peri044) [reply core/conversion/converters/impl/element_wise.cpp:147]: hardtanh looks a bit different [https://pytorch.org/docs/stable/generated/torch.nn.Hardtanh.html](url)
hardtanh compares the elements to -1 and 1 and replaces with default values (thresholds) where as clamp compares the input elements to thresholds directly. They both probably can match in certain scenarios

File: core/conversion/converters/impl/element_wise.cpp (hunk 2 above)
Comment (narendasan) [reply core/conversion/converters/impl/element_wise.cpp:152]: I see

File: core/conversion/converters/impl/element_wise.cpp (hunk 1 above)
Comment (narendasan) [reply core/conversion/converters/impl/element_wise.cpp:147]: hardtanh lets you configure the min and max values. really my point is we should be able to lower clamp and hardtanh to the same clip activation converter. This is something we could leave for later though

Comment (narendasan) [reply core/conversion/converters/impl/element_wise.cpp:147]: Ideally our converter library should be as small as possible and we should do as much as we can through lowering

**Review by narendasan — APPROVED** [review]

**Comment by narendasan** [comment]
@peri044 can you resolve merge conflicts? Ill merge after that.

**Comment by peri044** [comment]
Resolved conflicts

**Pull request merged**